core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,398 @@
+      } else {
+        responseCallback(error)
+      }
+    }
+  }
+
+  def handleTxnImmigration(transactionStateTopicPartitionId: Int, coordinatorEpoch: Int) {
+    txnManager.loadTransactionsForPartition(transactionStateTopicPartitionId, coordinatorEpoch)
+  }
+
+  def handleTxnEmigration(transactionStateTopicPartitionId: Int) {
+    txnManager.removeTransactionsForPartition(transactionStateTopicPartitionId)
+    txnMarkerChannelManager.removeStateForPartition(transactionStateTopicPartitionId)
+  }
+
+  def handleEndTransaction(transactionalId: String,
+                           pid: Long,
+                           epoch: Short,
+                           command: TransactionResult,
+                           responseCallback: EndTxnCallback): Unit = {
+
+    val errors = validateTransactionalId(transactionalId)
+    if (errors != Errors.NONE)
+      responseCallback(errors)
+    else
+      txnManager.getTransactionState(transactionalId) match {
+        case None =>
+          responseCallback(Errors.INVALID_PID_MAPPING)
+        case Some(metadata) =>
+          metadata synchronized {
+            if (metadata.pid != pid)
+              responseCallback(Errors.INVALID_PID_MAPPING)
+            else if (metadata.producerEpoch != epoch)
+              responseCallback(Errors.INVALID_PRODUCER_EPOCH)
+            else metadata.state match {
+              case Ongoing =>
+                commitOrAbort(transactionalId, pid, epoch, command, responseCallback, metadata)
+              case CompleteCommit =>
+                if (command == TransactionResult.COMMIT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case CompleteAbort =>
+                if (command == TransactionResult.ABORT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case _ =>
+                responseCallback(Errors.INVALID_TXN_STATE)
+            }
+          }
+      }
+  }
+
+  private def commitOrAbort(transactionalId: String,
+                            pid: Long,
+                            epoch: Short,
+                            command: TransactionResult,
+                            responseCallback: EndTxnCallback,
+                            metadata: TransactionMetadata) = {
+    val nextState = if (command == TransactionResult.COMMIT) PrepareCommit else PrepareAbort
+    val newMetadata = new TransactionMetadata(pid,
+      epoch,
+      metadata.txnTimeoutMs,
+      nextState,
+      metadata.topicPartitions,
+      metadata.transactionStartTime,
+      time.milliseconds())
+    metadata.prepareTransitionTo(nextState)
+
+    def logAppendCallback(errors: Errors): Unit = {
+      // we can respond to the client immediately and continue to write the txn markers if
+      // the log append was successful
+      responseCallback(errors)
```

Comment:
Hmm, is it? handleInitPid() seems to just call appendMetadataToLog(). There is no guarantee that abort/commit marker from the previous transaction has been sent by the inter broker thread.